docs/spokeclusterhealthcheck.md

```diff
@@ -0,0 +1,25 @@
+# Spoke cluster health check
+
+This document describes the hub cluster how to check the spoke cluster health, a new condition
+`SpokeClusterConditionAPIServerAvailable` will be introduced to describe whether the
+kube-apiserver of a spoke cluster is available
+
+Actors
+
+- agent on spoke cluster
+- hub controller on hub cluster
+
+## Health check
+
+1. after a spoke cluster is accepted by hub cluster admin, the agent on spoke cluster will
+create a [`Lease`](https://github.com/kubernetes/api/blob/master/coordination/v1/types.go#L27)
+on its namespace on hub cluster, then the agent will periodically (by default 10s)
```

Comment:
I think we need to talk about scale and how you plan to manage this.  If you have 5000 spokes, what you're describing here would be 500 qps doing nothing but heartbeating.  As with nodes, this presents a serious problem. The upstream KEP is here: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/0009-node-heartbeat.md .  A smaller resource (like coordination) helps to some degree because you aren't fanning out every watcher with large objects, but that is a lot of QPS for an idle system.

If you want to do this, I think it's important to describe what it will be used to do, how it is better or different than metrics, and then provide some meaningful way to manage the scale, QPS, and managing/changing update frequency from the hub to attempt to prevent overrun.

Reply:
+1 to David's comments.

> provide some meaningful way to manage the scale, QPS, and managing/changing update frequency from the hub to attempt to prevent overrun.

A knob is critical for this, also using Jitter on the frequency of the update. We should be able to tell users how to plan for a specific QPS number.

@deads2k what do you think is a reasonable qps to allocate for heartbeating?

Reply:
@deads2k and @pmorie, sorry for late

for QPS, if we have a reasonable QPS (or we can allow user set it), I think we can use `Lease.LeaseDurationSeconds` to coordinate the QPS

1. hub controller gets the count of spoke clusters and calculate the average lease duration time with a reasonable QPS and applies it to every spoke cluster, e.g. if the expected QPS is 100 and there are 5000 clusters, we set the `Lease.LeaseDurationSeconds` to 50s
2. spoke agent calculates its lease renew interval time by `Lease.LeaseDurationSeconds`, refer to kubelet implementation, we give a renew interval fraction (e.g. 0.25), we use `Lease.LeaseDurationSeconds * Fraction` to get the renew interval time, for above example, the renew interval time will be 12.5s

> how it is better or different than metrics

for migration case, we want to automatically migrate the workloads among the spoke clusters when some spoke clusters are unavailable,  if we use metrics, the hub controller need periodically poll all of spoke clusters to know whether a spoke cluster is available, I think this is an inefficient way, if we have a status to present whether a spoke cluster is available, the hub controller can list them on hub cluster, I think this is more efficient (except the spoke cluster api health, we may need to know the status of other spoke agents, so we may change the `SpokeClusterConditionAPIServerAvailable` to `SpokeClusterConditionAvailable`)

What's your opinions?

Reply:
should we consider adding a `LeaseDuration`  field in SpokeCluster API? So cluster-admin on hub is able to change that, and spoke agent can adjust the lease accordingly.

I think we need the condition on spokecluster also because reporting metrics will require additional request path from hub to spoke (or spoke to hub). With this condition, we can at least know whether registration/work agent is running with the minimum deployment.